board/piksiv3/patches/linux/0006-fpga-manager-custom-image-loader.patch

```diff
@@ -0,0 +1,228 @@
+@@ -69,7 +69,7 @@ static int fpga_mgr_write_init_sg(struct fpga_manager *mgr,
+ 	char *buf;
+ 	int ret;
+ 
+-	if (!mgr->mops->initial_header_size)
++	if (!info->initial_header_size)
+ 		return fpga_mgr_write_init_buf(mgr, info, NULL, 0);
+ 
+ 	/*
+@@ -78,7 +78,7 @@ static int fpga_mgr_write_init_sg(struct fpga_manager *mgr,
+ 	 */
+ 	sg_miter_start(&miter, sgt->sgl, sgt->nents, SG_MITER_FROM_SG);
+ 	if (sg_miter_next(&miter) &&
+-	    miter.length >= mgr->mops->initial_header_size) {
++	    miter.length >= info->initial_header_size) {
+ 		ret = fpga_mgr_write_init_buf(mgr, info, miter.addr,
+ 					      miter.length);
+ 		sg_miter_stop(&miter);
+@@ -87,12 +87,12 @@ static int fpga_mgr_write_init_sg(struct fpga_manager *mgr,
+ 	sg_miter_stop(&miter);
+ 
+ 	/* Otherwise copy the fragments into temporary memory. */
+-	buf = kmalloc(mgr->mops->initial_header_size, GFP_KERNEL);
++	buf = kmalloc(info->initial_header_size, GFP_KERNEL);
+ 	if (!buf)
+ 		return -ENOMEM;
+ 
+ 	len = sg_copy_to_buffer(sgt->sgl, sgt->nents, buf,
+-				mgr->mops->initial_header_size);
++				info->initial_header_size);
+ 	ret = fpga_mgr_write_init_buf(mgr, info, buf, len);
+ 
+ 	kfree(buf);
+@@ -295,7 +295,7 @@ int fpga_mgr_firmware_load(struct fpga_manager *mgr,
+ 
+ 	mgr->state = FPGA_MGR_STATE_FIRMWARE_REQ;
+ 
+-	ret = request_firmware(&fw, image_name, dev);
++    ret = request_firmware(&fw, image_name, dev);
+ 	if (ret) {
+ 		mgr->state = FPGA_MGR_STATE_FIRMWARE_REQ_ERR;
+ 		dev_err(dev, "Error requesting firmware %s\n", image_name);
+@@ -366,6 +366,7 @@ static ssize_t firmware_store(struct device *dev,
+ 
+ 	/* flags indicates whether to do full or partial reconfiguration */
+ 	info.flags = 0;
++    info.initial_header_size = mgr->mops->initial_header_size;
+ 
+ 	/* lose terminating \n */
+ 	strcpy(image_name, buf);
+@@ -373,7 +374,10 @@ static ssize_t firmware_store(struct device *dev,
+ 	if (image_name[len - 1] == '\n')
+ 		image_name[len - 1] = 0;
+ 
+-	ret = fpga_mgr_firmware_load(mgr, &info, image_name);
++    if (mgr->mops->load_firmware)
++        ret = mgr->mops->load_firmware(mgr, &info, image_name);
++    else
++	    ret = fpga_mgr_firmware_load(mgr, &info, image_name);
+ 	if (ret)
+ 		return ret;
+ 
+diff --git a/drivers/fpga/zynq-fpga.c b/drivers/fpga/zynq-fpga.c
+index 70b15b303471..9687b3817b83 100644
+--- a/drivers/fpga/zynq-fpga.c
++++ b/drivers/fpga/zynq-fpga.c
+@@ -19,6 +19,7 @@
+ #include <linux/completion.h>
+ #include <linux/delay.h>
+ #include <linux/dma-mapping.h>
++#include <linux/firmware.h>
+ #include <linux/fpga/fpga-mgr.h>
+ #include <linux/interrupt.h>
+ #include <linux/io.h>
+@@ -546,12 +547,92 @@ static enum fpga_mgr_states zynq_fpga_ops_state(struct fpga_manager *mgr)
+ 	return FPGA_MGR_STATE_UNKNOWN;
+ }
+ 
++int find_sync(const uint8_t *data, size_t size, size_t *off, bool *swapped)
++{
++    size_t i;
++    for (i = 0; i < size - 4; i++)
++    {
++        if (memcmp(data + i, "\x66\x55\x99\xAA", 4) == 0)
```

Comment:
What are these constants?  Can we add a couple #define's for them so they aren't magic numbers/strings?

Reply:
Done